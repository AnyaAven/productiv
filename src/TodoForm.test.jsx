import { test, expect, describe, beforeAll, afterAll, vi } from "vitest";
import { render, fireEvent } from "@testing-library/react";
import TodoForm from "./TodoForm.jsx";
import { TODOS, testFunction } from "./_testCommon.js";
import * as testCommon from "./_testCommon.js";

beforeAll(function () {
  vi.spyOn(testCommon, "testFunction");
});

afterAll(function () {
  vi.restoreAllMocks();
});

describe("smoke tests", function () {
  test("renders without crashing", function () {
    render(<TodoForm />);
  });
  test("matches snapshot on initial render", function () {
    const { container } = render(<TodoForm />);
    expect(container).toMatchSnapshot();
  });
});

describe("works", function () {
  test("updating the Todo item", function () {
    testCommon.testFunction.mockReturnValue();
    const { container } = render(
      <TodoForm
        initialFormData={TODOS[0]}
        handleSave={testFunction}
      />
    );

    fireEvent.input(description, { target: { value: "spinach" } });

    expect(testFunction).toHaveBeenCalledTimes(0);
    //User interacts with form
    fireEvent.click(container.querySelector(".TodoForm-Button"));
    //Test that the cb sent to the form is called,
    expect(testFunction).toHaveBeenCalledTimes(1);
  });

  test("creating a Todo item", function () {
    testCommon.testFunction.mockReturnValue();
    const { container } = render(
      <TodoForm
        initialFormData={{ title: "", priority: "", description: "" }}
        handleSave={testFunction}
      />
    );
    fireEvent.input(title, { target: { value: "Sleep" } });
    fireEvent.input(priority, { target: { value: 1 } });
    fireEvent.input(description, { target: { value: "sleep well" } });

    expect(testFunction).toHaveBeenCalledTimes(0);
    //User interacts with form
    fireEvent.click(container.querySelector(".TodoForm-Button"));
    //Test that the cb sent to the form is called,
    expect(testFunction).toHaveBeenCalledTimes(1);
  });

  test("handleChange callback works", function () {
    //TODO: how do write a test for handleChange ??
  });

});

