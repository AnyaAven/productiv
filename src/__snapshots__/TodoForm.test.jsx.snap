// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`smoke tests > matches snapshot on initial render 1`] = `
<div>
  <form
    class="TodoForm"
  >
    <div
      class="mb-3"
    >
      <input
        aria-label="Title"
        class="form-control"
        name="title"
        placeholder="Title"
        value="FIXME"
      />
    </div>
    <div
      class="mb-3"
    >
      <textarea
        aria-label="Description"
        class="form-control TodoForm-description"
        name="description"
        placeholder="Description"
      >
        FIXME
      </textarea>
    </div>
    <div
      class="mb-3"
    >
      <div
        class="w-75 d-flex justify-content-between"
      >
        <label
          class="d-inline-flex"
          for="TodoForm-priority"
        >
          Priority:  
        </label>
        <select
          class="form-control form-control-sm d-inline-flex TodoForm-priority"
          id="TodoForm-priority"
          name="priority"
        >
          <option
            value="1"
          >
            Ultra-Über
          </option>
          <option
            value="2"
          >
            Über
          </option>
          <option
            value="3"
          >
            Meh
          </option>
        </select>
      </div>
      <button
        class="btn-primary btn btn-sm TodoForm-addBtn"
      >
        Gø!
      </button>
    </div>
  </form>
</div>
`;
